### DISCUSSION TITLE
Could gpioCfgClock be made available

### DISCUSSION THREAD

**Discussion by pietdewinter** [Pi4J/pi4j#99] [General]

**Comment by hagego** [comment]
see also https://github.com/Pi4J/pi4j-v2/discussions/174

**Comment by savageautomate** [comment]
It looks like the Pi4J PIGPIO library does expose the `gpioCfgClock` function directly in the JNI native library.  So in theory you should be able to access this function via Java.  

https://github.com/Pi4J/pi4j-v2/blob/9763aa3ecbd8564890b1af7e9e367f74289bad39/libraries/pi4j-library-pigpio/src/main/java/com/pi4j/library/pigpio/internal/PIGPIO.java#L1272

However, it's not exposed via the higher level Pi4J framework or APIs.    So you would need to reference the JNI library directly in your code to execute this function.

**Reply by hagego** [reply]
I have tried it, but get a crash so far. According to pigpio documentation, this function must be called before gpioInitialise(), so to keep it simple for a first experiment  I hacked  PgioNativeImpl.gpioInitialise()
https://github.com/Pi4J/pi4j-v2/blob/dfbaa0ffcc57f187b589d2f0c66c99dcadb05c13/libraries/pi4j-library-pigpio/src/main/java/com/pi4j/library/pigpio/impl/PiGpioNativeImpl.java#L84
like this:

        if(!this.initialized) {
            // disable socket and pipes interfaces
            int rslt = PIGPIO.gpioCfgInterfaces(3);
            
            PIGPIO.gpioCfgClock(5,0,0);            

            // initialize the PiGpio native library
            result = PIGPIO.gpioInitialise();
            validateResult(result);
            
            // initialization successful
            this.initialized = true;
            logger.debug("[INITIALIZE] -- INITIALIZED SUCCESSFULLY");
        }

But get a crash:

2022-01-30 10:30:18 FINEST  com.pi4j.library.pigpio.impl.PiGpioNativeImpl [INITIALIZE] -> STARTED
2022-01-30 10:30:18 FINE    com.pi4j.library.pigpio.util.NativeLibraryLoader Attempting to load library [libpi4j-pigpio.so] using path: [/lib/armhf/libpi4j-pigpio.so]
2022-01-30 10:30:19 FINE    com.pi4j.library.pigpio.util.NativeLibraryLoader Library [libpi4j-pigpio.so] loaded successfully using embedded resource file: [/lib/armhf/libpi4j-pigpio.so]
2022-01-30 10:30:19 SEVERE  alarmpi.AlarmPi                Uncaught runtime exception in main thread: 'int com.pi4j.library.pigpio.internal.PIGPIO.gpioCfgClock(int, int, int)'
2022-01-30 10:30:19 SEVERE  alarmpi.AlarmPi                Uncaught runtime exception in main thread: null
2022-01-30 10:30:19 SEVERE  alarmpi.AlarmPi                com.pi4j.library.pigpio.internal.PIGPIO.gpioCfgClock(Native Method)
2022-01-30 10:30:19 SEVERE  alarmpi.AlarmPi                com.pi4j.library.pigpio.impl.PiGpioNativeImpl.gpioInitialise(PiGpioNativeImpl.java:93)
2022-01-30 10:30:19 SEVERE  alarmpi.AlarmPi                com.pi4j.library.pigpio.PiGpio.initialise(PiGpio.java:174)


Is there anything else that needs to be done before a JNI function can be used ? Sorry, no expert in JNI...

**Reply by hagego** [reply]
I digged further and found that this function is commented out in the wrapper header:
https://github.com/Pi4J/pi4j-v2/blob/dfbaa0ffcc57f187b589d2f0c66c99dcadb05c13/libraries/pi4j-library-pigpio/src/main/native/com_pi4j_library_pigpio_internal_PIGPIO.h#L898

And does not exist in the .c file. I tried to fix/hack also this part, but had to give up for now as I ran into too many issues I could not solve with my limited understanding of the structure of the project

**Reply by hagego** [reply]
As soon as I compile the native code (using release 2.1.1 sources w/o any changes) using -Pnative in mvn, the JVM crashes when loading libpi4j-pigpio.so from the resulting jar. I guess cross-compiling the native code should work, or are there any known issues ? So something wrong with my setup, for the moment I got  stuck

**Reply by hagego** [reply]
I found out that the crash happened because I was using an arm6 Pi which is not supported by the toolchain. Using the docker based toolchain things are now working.
I can confirm that my sound output is working fine again when calling gpioCfgClock(5,0,0) gefore Initialise().
The change is simple, requires around 10 lines of code, I hope this can be made part of the master branch

**Comment by FDelporte** [comment]
@hagego that is really great news and thanks a lot for taking the time to fully research this topic!!!
would you be willing to provide a pull request with this change?

**Reply by hagego** [reply]
willing - sure ;-)
I am no git expert but I should manage. But I expect someone has to grant me permissions to do so ?

**Reply by FDelporte** [reply]
two possible solutions:

1. Quick - not good: you send me the changes and I create a pull request, but that would make me the contributor ;-)
2. Not quick - very good: go to the top of this page, push the fork button which makes a copy of this project in your account, modify there, and from your repository create a pull request back to the Pi4J project. 

Or try 2. and if it fails, switch to 1. :-)

**Reply by hagego** [reply]
Done. At least I tried to do so ;-)
PiGpio interface got a new method gpioCfgClock()

**Reply by FDelporte** [reply]
This is now in release 2.2.0, see https://pi4j.com/about/release-notes/
